File: app/src/main/java/tag_them/metoothanks/items/Image.java
package tag_them.metoothanks.items;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.drawable.BitmapDrawable;
import android.view.Gravity;

import tag_them.metoothanks.CanvasView;
import tag_them.metoothanks.R;

public class Image extends Item {
	private BitmapDrawable bitmapDrawable;
	
	public Image(Bitmap bitmap, CanvasView hostView) {
		super(hostView, bitmap.getWidth(), bitmap.getHeight());
		bitmapDrawable = new BitmapDrawable(hostView.getContext().getResources(), bitmap);
		bitmapDrawable.setBounds(0, 0, right, bottom);
		bitmapDrawable.setGravity(Gravity.FILL);
	}
	
	@Override
	public void draw(Canvas canvas) {
		getPaint().setAlpha(255);
		bitmapDrawable.setBounds(left, top, right, bottom);
		bitmapDrawable.draw(canvas);
		drawFrame(canvas);
	}
	
	@Override
	public int getMenuID() {
		return R.menu.item_image_menu;
	}
	
	@Override
	public CanvasView.ClickListener getClickListener() {
		return item -> {
			switch (item.getItemId()) {
				case R.id.action_image_rotate_counterclockwise:
					rotateImage(-90);
					return true;
					
				case R.id.action_image_rotate_clockwise:
					rotateImage(90);
					return true;
			}
			return true;
		};
	}
	
	private void rotateImage(float degrees) {
		Matrix matrix = new Matrix();
		matrix.postRotate(degrees);
		
		Bitmap bitmap = Bitmap.createBitmap(bitmapDrawable.getBitmap(), 0, 0,
			bitmapDrawable.getIntrinsicWidth(), bitmapDrawable.getIntrinsicHeight(), matrix, true);
		
		swapWidthHeight();
		
		bitmapDrawable = new BitmapDrawable(hostView.getResources(), bitmap);
		hostView.postInvalidate();
	}
	
	private void swapWidthHeight() {
		int tmp = right;
		right = bottom;
		bottom = tmp;
	}
}

File: app/src/main/java/org/tag_them/metoothanks/items/Image.kt
package org.tag_them.metoothanks.items

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.drawable.BitmapDrawable
import android.view.Gravity
import android.view.MenuItem
import org.tag_them.metoothanks.CanvasView
import org.tag_them.metoothanks.R


class Image(bitmap: Bitmap, hostView: CanvasView) : Item(hostView, bitmap.width, bitmap.height) {
	override fun handleMenuItemClick(item: MenuItem): Boolean {
		when (item.itemId) {
			R.id.action_image_rotate_clockwise        -> rotateImage(90f)
			R.id.action_image_rotate_counterclockwise -> rotateImage(-90f)
			else                                      -> return false
		}
		
		hostView.postInvalidate()
		
		return true
	}
	
	override val item_menu_id: Int = R.menu.item_image_menu
	
	var bitmapDrawable = BitmapDrawable(hostView.context.resources, bitmap).apply {
		setBounds(0, 0, right, bottom)
		gravity = Gravity.FILL
	}
	
	override fun draw(canvas: Canvas) {
		paint.alpha = 255
		
		with(bitmapDrawable) {
			setBounds(left, top, right, bottom)
			draw(canvas)
		}
		hostView.postInvalidate()
	}
	
	private fun rotateImage(angle: Float) {
		swapWidthHeight()
		bitmapDrawable = BitmapDrawable(hostView.resources, rotate(bitmapDrawable.bitmap, angle))
		hostView.postInvalidate()
	}
	
	private fun swapWidthHeight() {
		print("before: $width $height $bounds")
		val tmp = height
		bottom = top + width
		right = left + tmp
		println(", after: $width $height $bounds")
	}
	
	fun rotate(source: Bitmap, angle: Float): Bitmap {
		val matrix = Matrix()
		matrix.postRotate(angle)
		return Bitmap.createBitmap(source, 0, 0, source.width, source.height, matrix, true)
	}
}

